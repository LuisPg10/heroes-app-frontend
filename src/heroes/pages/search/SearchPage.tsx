import { useSearchParams } from 'react-router';
import { useQuery } from '@tanstack/react-query';

import { searchHeroesAction } from '@/heroes/actions/search-heroes.action';

import { CustomJumbotron } from '@/components/custom/CustomJumbotron';
import { HeroStats } from '@/heroes/components/HeroStats';
import { SearchControls } from './ui/SearchControls';
import { CustomBreadcrumbs } from '@/components/custom/CustomBreadcrumbs';
import { HeroGrid } from '@/heroes/components/HeroGrid';

export const SearchPage = () => {
  const [searchParams] = useSearchParams();

  const name = searchParams.get('name') ?? undefined;
  const strength = searchParams.get('strength') ?? undefined;

  const { data: heroes } = useQuery({
    queryKey: ['heroes/search', { name, strength }],
    queryFn: () => searchHeroesAction({ name, strength }),
    staleTime: 1000 * 60 * 5, // 5 minutos
  });

  return (
    <>
      <CustomJumbotron
        title="Búsqueda de SuperHéroes"
        description="Descubre, explora y administra superhéroes y villanos"
      />

      {/* Breadcrumbs */}
      <CustomBreadcrumbs currentPage="Buscador de héroes" />

      {/* Stats Dashboard */}
      <HeroStats />

      {/* Filter and Search */}
      <SearchControls />

      {/* List of heroes */}
      <HeroGrid heroes={heroes} />
    </>
  );
};

export default SearchPage;
