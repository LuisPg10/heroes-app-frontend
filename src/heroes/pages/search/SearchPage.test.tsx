import { render, screen, waitFor } from '@testing-library/react';
import { beforeEach, describe, expect, test, vi } from 'vitest';
import { MemoryRouter } from 'react-router';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { searchHeroesAction } from '@/heroes/actions/search-heroes.action';
import SearchPage from './SearchPage';
import type { Hero } from '@/heroes/types/hero.interface';

vi.mock('@/heroes/actions/search-heroes.action');
const mockSearchHeroesAction = vi.mocked(searchHeroesAction);

vi.mock('@/components/custom/CustomJumbotron', () => ({
  CustomJumbotron: () => <div>CustomJumbotron</div>,
}));

vi.mock('@/components/custom/CustomBreadcrumbs', () => ({
  CustomBreadcrumbs: () => <div>CustomBreadcrumbs</div>,
}));

vi.mock('./ui/SearchControls', () => ({
  SearchControls: () => <div>SearchControls</div>,
}));

vi.mock('@/heroes/components/HeroGrid', () => ({
  HeroGrid: ({ heroes = [] }: { heroes: Hero[] }) => (
    <div data-testid="hero-grid">
      {heroes.map((hero) => (
        <p key={hero.id}>{hero.name}</p>
      ))}
    </div>
  ),
}));

const queryCient = new QueryClient();

const renderSearchPage = (initialEntries?: string[]) => {
  return render(
    <MemoryRouter initialEntries={initialEntries}>
      <QueryClientProvider client={queryCient}>
        <SearchPage />
      </QueryClientProvider>
    </MemoryRouter>
  );
};

describe('SearchPage', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  test('should render SearchPage with defautl values', () => {
    renderSearchPage();

    expect(mockSearchHeroesAction).toHaveBeenCalledWith({
      name: undefined,
      strength: undefined,
    });
  });

  test('should match to snapshot', () => {
    const { container } = renderSearchPage();

    expect(container).toMatchSnapshot();
  });

  test('should call search action with name parameter', () => {
    renderSearchPage(['/search?name=superman']);

    expect(mockSearchHeroesAction).toHaveBeenCalledWith({
      name: 'superman',
      strength: undefined,
    });
  });

  test('should call search action with strength parameter', () => {
    renderSearchPage(['/search?strength=6']);

    expect(mockSearchHeroesAction).toHaveBeenCalledWith({
      name: undefined,
      strength: '6',
    });
  });

  test('should call search action with name and strength parameter', () => {
    renderSearchPage(['/search?name=batman&strength=8']);

    expect(mockSearchHeroesAction).toHaveBeenCalledWith({
      name: 'batman',
      strength: '8',
    });
  });

  test('should render HeroGrid with search results', async () => {
    const mockHeroes = [
      { id: '1', name: 'Clark Kent' } as unknown as Hero,
      { id: '2', name: 'Bruce Wayne' } as unknown as Hero,
    ];

    mockSearchHeroesAction.mockResolvedValue(mockHeroes);

    renderSearchPage();

    await waitFor(() => {
      expect(screen.getByText('Clark Kent')).toBeDefined();
      expect(screen.getByText('Bruce Wayne')).toBeDefined();
    });
  });
});
