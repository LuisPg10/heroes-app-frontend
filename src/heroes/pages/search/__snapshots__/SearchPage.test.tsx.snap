// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchPage > should match to snapshot 1`] = `
<div>
  <div>
    CustomJumbotron
  </div>
  <div>
    CustomBreadcrumbs
  </div>
  <div>
    Cargando...
  </div>
  <div>
    SearchControls
  </div>
  <div
    data-testid="hero-grid"
  />
</div>
`;
